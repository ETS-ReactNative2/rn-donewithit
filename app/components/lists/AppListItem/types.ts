import {
  Animated,
  GestureResponderEvent,
  ImageSourcePropType,
  StyleProp,
  ViewStyle
} from 'react-native'

export type AppListItemProps = {
  title: string
  subTitle?: string
  titleMaxLines?: number
  subTitleMaxLines?: number
  image: ImageSourcePropType
  style?: StyleProp<ViewStyle>
  onPress?: (e?: GestureResponderEvent) => void
  renderRightActions?: (
    progressAnimatedValue: Animated.AnimatedInterpolation,
    dragAnimatedValue: Animated.AnimatedInterpolation
  ) => React.ReactNode
}
