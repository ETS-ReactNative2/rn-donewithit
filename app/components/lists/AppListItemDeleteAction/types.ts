export type AppListItemDeleteActionProps = {
  onDelete?: () => void
}
