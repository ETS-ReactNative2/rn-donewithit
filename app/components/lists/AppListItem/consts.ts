const CONSTS = {
  SUBTITLE_FONT_SIZE: 18,
  MARGIN_RIGHT: 10,
  AVATAR_SIZE: 80,
  ICON_SIZE: 20,
  TITLE_MAX_LINES: 1,
  SUBTITLE_MAX_LINES: 3,
  PADDING: 5
}

export default CONSTS
