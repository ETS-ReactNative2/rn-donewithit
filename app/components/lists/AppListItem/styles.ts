import { StyleSheet } from 'react-native'
import CONSTS from './consts'
import COLORS from '../../../config/colors'

const styles = StyleSheet.create({
  listItem: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  avatarWrap: {
    height: '100%'
  },
  avatar: {
    width: CONSTS.AVATAR_SIZE,
    height: CONSTS.AVATAR_SIZE,
    borderRadius: CONSTS.AVATAR_SIZE / 2,
    marginRight: CONSTS.MARGIN_RIGHT
  },
  infoBox: {
    flex: 1,
    height: '100%'
  },
  title: {
    fontSize: CONSTS.SUBTITLE_FONT_SIZE
  },
  subTitle: {
    color: COLORS.GRAY,
    marginTop: CONSTS.PADDING
  }
})

export default styles
