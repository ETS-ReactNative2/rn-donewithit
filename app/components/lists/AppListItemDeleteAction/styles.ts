import { StyleSheet } from 'react-native'
import COLORS from '../../../config/colors'

const styles = StyleSheet.create({
  container: {
    width: 80,
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: COLORS.DANGER
  }
})

export default styles
