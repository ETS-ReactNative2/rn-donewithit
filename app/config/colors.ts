const COLORS = {
  PRIMARY: 'tomato',
  SECONDARY: 'turquoise',
  BACKGROUND_COLOR: '#f8f4f4',
  INPUT_BACKGROUND_COLOR: '#f9f2f2',
  DANGER: '#ff5252',
  GRAY: '#6e6969',
  LIGHT_GRAY: '#f8f4f4',
  WHITE: '#fff',
  BLACK: '#000',
  SILVER: '#C0C0C0',
  YELLOW: '#ffe66d'
}

export default COLORS
