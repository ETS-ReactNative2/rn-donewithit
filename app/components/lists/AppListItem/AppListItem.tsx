import React, { useCallback } from 'react'
import {
  View,
  Image,
  GestureResponderEvent,
  TouchableHighlight,
  Animated
} from 'react-native'
import AppText from '../../AppText'
import COLORS from '../../../config/colors'
import styles from './styles'
import { AppListItemProps } from './types'
import { GestureHandlerRootView, Swipeable } from 'react-native-gesture-handler'
import AppIcon from '../../AppIcon'
import CONSTS from './consts'
function AppListItem({
  title,
  subTitle,
  titleMaxLines = CONSTS.TITLE_MAX_LINES,
  subTitleMaxLines = CONSTS.SUBTITLE_MAX_LINES,
  image,
  style,
  onPress,
  renderRightActions
}: AppListItemProps) {
  const handlePress = useCallback((e?: GestureResponderEvent) => {
    typeof onPress === 'function' && onPress(e)
  }, [])

  const handleRenderRightActions = useCallback(
    (
      progressAnimatedValue: Animated.AnimatedInterpolation,
      dragAnimatedValue: Animated.AnimatedInterpolation
    ) => {
      if (typeof renderRightActions === 'function') {
        return renderRightActions(progressAnimatedValue, dragAnimatedValue)
      }
      return null
    },
    []
  )

  return (
    <GestureHandlerRootView>
      <Swipeable renderRightActions={handleRenderRightActions}>
        <TouchableHighlight
          underlayColor={COLORS.LIGHT_GRAY}
          activeOpacity={0.85}
          onPress={(e?: GestureResponderEvent) => {
            handlePress(e)
          }}
        >
          <View style={[style, styles.listItem]}>
            <View style={styles.avatarWrap}>
              <Image style={styles.avatar} source={image} />
            </View>
            <View style={styles.infoBox}>
              <AppText style={styles.title} numberOfLines={titleMaxLines}>
                {title}
              </AppText>
              {subTitle && (
                <AppText
                  style={[styles.title, styles.subTitle]}
                  numberOfLines={subTitleMaxLines}
                >
                  {subTitle}
                </AppText>
              )}
            </View>
            <AppIcon
              name="chevron-right"
              size={CONSTS.ICON_SIZE}
              color={COLORS.GRAY}
            />
          </View>
        </TouchableHighlight>
      </Swipeable>
    </GestureHandlerRootView>
  )
}

export default AppListItem
