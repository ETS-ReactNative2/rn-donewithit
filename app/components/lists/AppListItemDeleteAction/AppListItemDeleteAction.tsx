import React, { useCallback } from 'react'
import { View } from 'react-native'
import COLORS from '../../../config/colors'
import CONSTS from './consts'
import { MaterialCommunityIcons } from '@expo/vector-icons'
import { TouchableWithoutFeedback } from 'react-native-gesture-handler'
import { AppListItemDeleteActionProps } from './types'
import styles from './styles'

function AppListItemDeleteAction({ onDelete }: AppListItemDeleteActionProps) {
  const handlePress = useCallback(() => {
    typeof onDelete === 'function' && onDelete()
  }, [])

  return (
    <TouchableWithoutFeedback onPress={handlePress}>
      <View style={styles.container}>
        <MaterialCommunityIcons
          name="trash-can"
          size={CONSTS.ICON_SIZE}
          color={COLORS.WHITE}
        />
      </View>
    </TouchableWithoutFeedback>
  )
}

export default AppListItemDeleteAction
