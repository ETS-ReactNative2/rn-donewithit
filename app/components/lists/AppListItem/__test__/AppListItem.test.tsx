import React from 'react'
import renderer from 'react-test-renderer'
import AppListItem from '../index'
import { MOSH } from '../../../../config/images'
import { Animated, ImageSourcePropType, StyleSheet } from 'react-native'
import AppListItemDeleteAction from '../../AppListItemDeleteAction'

jest.mock(
  '@expo/vector-icons/build/vendor/react-native-vector-icons/lib/create-icon-set.js',
  () => {
    return () => ''
  }
)
type Message = {
  id: number
  title: string
  description: string
  image: ImageSourcePropType
}
const msg: Message = {
  id: 1,
  title: 'Mosh Hamedani',
  description: 'Hey! Is this item still available?',
  image: MOSH
}

const handleRenderRightActions = (
  progress: Animated.AnimatedInterpolation,
  dragX: Animated.AnimatedInterpolation,
  item: Message
): JSX.Element => {
  return (
    <AppListItemDeleteAction
      onDelete={() => {
        console.log(item.id)
      }}
    />
  )
}
const styles = StyleSheet.create({
  item: {
    padding: 20
  }
})

describe('AppListItem', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <AppListItem
          style={styles.item}
          title={msg.title}
          subTitle={msg.description}
          image={msg.image}
          renderRightActions={(
            progress: Animated.AnimatedInterpolation,
            dragX: Animated.AnimatedInterpolation
          ) => {
            return handleRenderRightActions(progress, dragX, msg)
          }}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
