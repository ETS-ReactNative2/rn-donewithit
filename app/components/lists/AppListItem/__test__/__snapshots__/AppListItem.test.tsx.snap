// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppListItem renders correctly 1`] = `
<View>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "flexDirection": "row-reverse",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateX": -10000,
            },
          ],
        }
      }
    >
      <RNGestureHandlerButton
        collapsable={false}
        exclusive={true}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
      >
        <View
          accessible={true}
          style={Object {}}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#ff5252",
                "height": "100%",
                "justifyContent": "center",
                "width": 80,
              }
            }
          >
            <Text />
          </View>
        </View>
      </RNGestureHandlerButton>
      <View
        onLayout={[Function]}
      />
    </View>
    <View
      collapsable={false}
      forwardedRef={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      pointerEvents="auto"
      style={
        Object {
          "transform": Array [
            Object {
              "translateX": -0,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Array [
              Object {
                "padding": 20,
              },
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "height": "100%",
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "borderRadius": 40,
                  "height": 80,
                  "marginRight": 10,
                  "width": 80,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "height": "100%",
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "Courier",
                    "fontSize": 13,
                  },
                  Object {
                    "fontSize": 18,
                  },
                ]
              }
            >
              Mosh Hamedani
            </Text>
            <Text
              numberOfLines={3}
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "Courier",
                    "fontSize": 13,
                  },
                  Array [
                    Object {
                      "fontSize": 18,
                    },
                    Object {
                      "color": "#6e6969",
                      "marginTop": 5,
                    },
                  ],
                ]
              }
            >
              Hey! Is this item still available?
            </Text>
          </View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</View>
`;
